### ISSUE TITLE
Preserve comments in YAML front matter

### ISSUE THREAD

**Issue by infotexture** [prettier/prettier#4390]
The fixes for #3321 in #3802 only work if the YAML front matter block actually contains variable content.

As noted in #4162:

> I noticed prettier will retain Jekyll-style front matter blocks if they have content but remove them if they are empty. I believe this to be incorrect behavior as Jekyll requires the front-matter block to be present to process the file, even if that block has no content. Specifically this is happening with a .scss file in Jekyll with empty front-matter.

Many Jekyll projects use comments in the YAML front matter for SCSS files to ensure the front matter is preserved and explain its purpose:

```yaml
---
# Keep this YAML front matter block to trigger Jekyll processing
# See https://jekyllrb.com/docs/frontmatter for more information
---
```

However, Prettier strips these comments, so the resulting front matter block is effectively empty and thus discarded.

Discussion in https://github.com/jonschlinkert/gray-matter/issues/65 and #4243 suggests efforts are underway to preserve empty front matter, but it would be helpful to retain any comments in the process.

**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBadAdKBiABAaTjgAdcYALASwGdcBNAQQFkAZXAMwCdoZcBbAIYx4nXACMANhDABrMhDKdKAc2VxRAKTgyAnhIm5i3MHGrVKUZVjwBlIrnLDi1JAHpXAK216JnMQDpIPlcAE2lqVy4eQWF1DghRPgS4XAt2BJjKaCx0VCwsAAFKPmIE3gByADcBJQFJU3KAbkLi0s4KsQFqOCaWkrLcDBAYHWI4IeaoIv723HKJAR0IAFcYXqnWgfLIWAQ1yZAAGhAIYhgsqGpkUBruAHcABRqEK5QBSohKEKOQMU4BWRwGA2YgAizKZAwTjLODHCzddoPf7KQTIdgCCTdY4eagADwAQv9AcCBHw4CwLHA0RisSBQZwEcgQNQwGYfkYLDAAOpfCjIAAcAAZjkYIN0uf9iEyjKZ1JUqcdOHAAI7LShKpECFECamY2HMuB8SiQ6H68yWCRwACKywg8BNMOOMDqPJCfKQACYnf9KBJwQBhCB8VEoKDQBUgZbdAAqdVe6L1xy6MF1tLCYFT+pWMGIqw9mYAvgWgA)
```sh
--parser scss
--single-quote
```

**Input:**
```scss
---
# Keep this YAML front matter block to trigger Jekyll processing
# See https://jekyllrb.com/docs/frontmatter for more information
---

@import 'variables';
@import 'base';
@import "type";
@import 'layout';
@import 'content';

```

**Output:**
```scss
@import 'variables';
@import 'base';
@import 'type';
@import 'layout';
@import 'content';

```

**Expected behavior:**

The YAML front matter block — _including comments_  — should be preserved:

```scss
---
# Keep this YAML front matter block to trigger Jekyll processing
# See https://jekyllrb.com/docs/frontmatter for more information
---

@import 'variables';
@import 'base';
@import 'type';
@import 'layout';
@import 'content';
```

**Comment by evilebottnawi** [comment]
@infotexture Feel free to be champion this issue :+1:

**Comment by infotexture** [comment]
@evilebottnawi Am I not (by reporting) its de-facto champion? 😄

**Comment by duailibe** [comment]
Do we even need `graymatter` for this? Looking at its source code, it's not doing anything relevant besides checking the text starts with `---` and ends with `---`

I guess we could simply print the `originalText` between these two delimiters

**Comment by evilebottnawi** [comment]
@infotexture i.e. feel free to send PR :smile:

**Comment by infotexture** [comment]
I'm afraid that's a bit beyond my ken. 😕 Best to leave that to more capable coders.

**Comment by duailibe** [comment]
No pressure @infotexture.. Thanks for the report! :)

**Comment by infotexture** [comment]
@eliasmeire Thanks for the quick fix in #4392. 👍
